### PR TITLE
Modules for Autograd

### PR DESCRIPTION
**Pull request by allanzelener**
This PR creates a modules folder for autograd and adds MaskedBatchNormalization, LayerNormalization, and SoftAttention modules.

These modules are mid-sized functions with parameters that can be used in larger models. Typically an equivalent "nn" module is not available or has some issue that makes it unsuitable for use with autograd. E.g., nn.BatchNormalization accumulates running mean and std estimates that would be frozen in optimized mode.

### CODE REVIEW

**Comment by alexbw** [comment]
The LessThan test is failing, but that has nothing to do with here, it's a separate issue. Merging.

**Comment by alexbw** [comment]
@allanzelener These tests are failing: `Modules_MaskedBatchNormalization`, `Modules_SoftAttention`, `Modules_LayerNormalization`. Please fix ASAP. Will have to revert otherwise.

```
Optimized_Modules_SoftAttention
 Function call failed
test/test.lua:1314: attempt to index field 'module' (a nil value)
stack traceback:
       	test/test.lua:1314: in function <test/test.lua:1313>
       	[C]: in function 'xpcall'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:477: in function '_pcall'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:436: in function '_run'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:355: in function 'run'
       	test/test.lua:1955: in main chunk
       	[C]: in function 'dofile'
       	...wiltschko/anaconda/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
       	[C]: in ?

--------------------------------------------------------------------------------
Optimized_SinhCoshTanh
incorrect gradient for sinh
BOOL violation condition=false
stack traceback:
       	test/test.lua:636: in function <test/test.lua:627>
--------------------------------------------------------------------------------
Optimized_SinhCoshTanh
incorrect gradient for cosh
BOOL violation condition=false
stack traceback:
       	test/test.lua:636: in function <test/test.lua:627>
--------------------------------------------------------------------------------
Optimized_Modules_LayerNormalization
 Function call failed
test/test.lua:1277: attempt to index field 'module' (a nil value)
stack traceback:
       	test/test.lua:1277: in function <test/test.lua:1276>
       	[C]: in function 'xpcall'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:477: in function '_pcall'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:436: in function '_run'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:355: in function 'run'
       	test/test.lua:1955: in main chunk
       	[C]: in function 'dofile'
       	...wiltschko/anaconda/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
       	[C]: in ?

--------------------------------------------------------------------------------
Optimized_Modules_MaskedBatchNormalization
 Function call failed
test/test.lua:1366: attempt to index field 'module' (a nil value)
stack traceback:
       	test/test.lua:1366: in function <test/test.lua:1365>
       	[C]: in function 'xpcall'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:477: in function '_pcall'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:436: in function '_run'
       	/Users/awiltschko/anaconda/share/lua/5.2/torch/Tester.lua:355: in function 'run'
       	test/test.lua:1955: in main chunk
       	[C]: in function 'dofile'
       	...wiltschko/anaconda/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
       	[C]: in ?
```

**Comment by allanzelener** [comment]
I can't reproduce test failures on master. It looks like your test is out
of sync with commits from August 4th.
